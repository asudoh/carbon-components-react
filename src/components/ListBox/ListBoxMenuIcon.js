import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';
import { iconCaretDown } from 'carbon-icons';
import Icon from '../Icon';

export const translationIds = {
  'close.menu': 'close.menu',
  'open.menu': 'open.menu',
};

const defaultTranslations = {
  [translationIds['close.menu']]: 'Close menu',
  [translationIds['open.menu']]: 'Open menu',
};

/**
 * `ListBoxMenuIcon` is used to orient the icon up or down depending on the
 * state of the menu for a given `ListBox`
 */
const ListBoxMenuIcon = ({ isOpen, prefix, translateWithId: t }) => {
  const className = cx({
    [`${prefix}--list-box__menu-icon`]: true,
    [`${prefix}--list-box__menu-icon--open`]: isOpen,
  });
  const description = isOpen ? t('close.menu') : t('open.menu');
  return (
    <div className={className}>
      <Icon icon={iconCaretDown} description={description} alt={description} />
    </div>
  );
};

ListBoxMenuIcon.propTypes = {
  /**
   * Specify whether the menu is currently open, which will influence the
   * direction of the menu icon
   */
  isOpen: PropTypes.bool.isRequired,

  /**
   * The selector prefix
   */
  prefix: PropTypes.string,

  /**
   * i18n hook used to provide the appropriate description for the given menu
   * icon. This function takes in an id defined in `translationIds` and should
   * return a string message for that given message id.
   */
  translateWithId: PropTypes.func.isRequired,
};

ListBoxMenuIcon.defaultProps = {
  prefix: 'bx',
  translateWithId: id => defaultTranslations[id],
};

export default ListBoxMenuIcon;
