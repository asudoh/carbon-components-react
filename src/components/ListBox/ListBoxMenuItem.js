import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';

/**
 * `ListBoxMenuItem` is a helper component for managing the container class
 * name, alongside any classes for any corresponding states, for a generic list
 * box menu item.
 */
const ListBoxMenuItem = ({
  children,
  isActive,
  isHighlighted,
  prefix,
  ...rest
}) => {
  const className = cx({
    [`${prefix}--list-box__menu-item`]: true,
    [`${prefix}--list-box__menu-item--active`]: isActive,
    [`${prefix}--list-box__menu-item--highlighted`]: isHighlighted,
  });
  return (
    <div className={className} {...rest}>
      {children}
    </div>
  );
};

ListBoxMenuItem.propTypes = {
  /**
   * Specify any children nodes that hsould be rendered inside of the ListBox
   * Menu Item
   */
  children: PropTypes.node,

  /**
   * Specify whether the current menu item is "active".
   */
  isActive: PropTypes.bool.isRequired,

  /**
   * Specify whether the current menu item is "highlighed".
   */
  isHighlighted: PropTypes.bool.isRequired,

  /**
   * The selector prefix.
   */
  prefix: PropTypes.string,
};

ListBoxMenuItem.defaultProps = {
  isActive: false,
  isHighlighted: false,
  prefix: 'bx',
};

export default ListBoxMenuItem;
