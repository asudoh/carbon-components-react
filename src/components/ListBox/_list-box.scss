@import 'carbon-components/scss/globals/scss/colors';
@import 'carbon-components/scss/globals/scss/import-once';
@import 'carbon-components/scss/globals/scss/css--helpers';
@import 'carbon-components/scss/globals/scss/vars';

$list-box-width: rem(300px);
$list-box-height: rem(40px);

@include exports('list-box') {
  .bx--list-box {
    background-color: $field-01;
    width: $list-box-width;
    height: $list-box-height;
  }

  .bx--list-box--inline {
    background-color: inherit;
  }

  .bx--list-box__field {
    display: flex;
    align-items: center;
    justify-content: space-between;
    height: $list-box-height;
    cursor: pointer;
    user-select: none;
    padding: 0 1rem;
    font-size: rem(14px);
    font-weight: 600;
    color: $text-01;
  }

  .bx--list-box__field:focus {
    outline: 1px solid $brand-01;
  }

  .bx--list-box__field:focus ~ .bx--list-box__menu {
    box-shadow: 0 1px 0 0 $brand-01, 1px 0 0 0 $brand-01, -1px 0 0 0 $brand-01;
  }

  .bx--list-box__field--inline {
    display: inline-flex;
    justify-content: normal;
    color: $brand-01;
  }

  .bx--list-box__field--inline:focus {
    outline: 1px solid $brand-01;
  }

  .bx--list-box__field--inline:focus ~ .bx--list-box__menu {
    outline: 1px solid $brand-01;
  }

  .bx--list-box__field--inline > .bx--list-box__label {
    margin-right: rem(10px);
  }

  .bx--list-box__field--inline.bx--list-box__field--open {
    outline: 1px solid transparent;
  }

  .bx--list-box__label {
    display: flex;
    align-items: center;
    height: 100%;
  }

  .bx--list-box__badge {
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;
    // TODO: sync with margin
    left: -0.25rem;
    width: rem(18px);
    height: rem(18px);
    // TODO: is this what we use for white?
    color: $ui-01;
    background-color: $brand-01;
    border-radius: 50%;
    // TODO: typography helpers?
    font-size: rem(12px);
    font-weight: 200;
    margin: 0.5em;
  }

  .bx--list-box__badge:focus {
    outline: 1px solid $brand-01;
    outline-offset: 0.25rem;
  }

  .bx--list-box__badge > .bx--list-box__clear {
    display: none;
    visibility: hidden;
  }

  .bx--list-box__badge:focus > .bx--list-box__clear,
  .bx--list-box__badge:hover > .bx--list-box__clear {
    display: block;
    visibility: visible;
  }

  .bx--list-box__badge:focus > .bx--list-box__badge-text,
  .bx--list-box__badge:hover > .bx--list-box__badge-text {
    display: none;
    visibility: hidden;
  }

  .bx--list-box__menu {
    @include box-shadow();

    position: absolute;
    left: 0;
    right: 0;
    top: $list-box-height;
    background-color: $ui-01;
    max-height: rem(120px);
    overflow-y: auto;
  }

  .bx--list-box__menu-item {
    display: flex;
    align-items: center;
    height: rem(40px);
    padding: 0 1rem;
    cursor: pointer;
    user-select: none;
  }

  .bx--list-box__menu-item:hover,
  .bx--list-box__menu-item--highlighted {
    background-color: rgba($field-01, 0.1);
  }

  .bx--list-box__menu-icon {
    fill: $brand-01;
    transition: transform 200ms $bx--standard-easing;
  }

  .bx--list-box__menu-icon--open {
    transform: rotate(180deg);
  }

  .bx--list-box__clear {
    display: block;
    width: rem(18px);
    height: rem(18px);
    fill: $brand-02;
    background-color: white;
    border-radius: 50%;
  }
}
