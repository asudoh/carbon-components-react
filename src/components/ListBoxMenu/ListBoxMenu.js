import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Icon from '../Icon';

const ListBoxMenu = ({
  items,
  selectedItem,
  highlightedIndex,
  itemToString,
  getItemProps,
}) => (
  <div className="bx--list-box__menu">
    {items.map((item, i) => (
      <div
        key={item.id}
        className={cx({
          'bx--list-box__menu-item': true,
          'bx--list-box__menu-item--highlighted': highlightedIndex === i,
          'bx--list-box__menu-item--active': selectedItem.indexOf(item) !== -1,
        })}
        {...getItemProps({
          item,
          index: i,
        })}>
        <input
          id={item.id}
          name={itemToString(item)}
          className="bx--checkbox"
          type="checkbox"
          checked={selectedItem.indexOf(item) !== -1}
          readOnly={true}
        />
        <label className="bx--checkbox-label">
          <span className="bx--checkbox-appearance">
            <Icon className="bx--checkbox-checkmark" name="checkmark" />
          </span>
          {itemToString(item)}
        </label>
      </div>
    ))}
  </div>
);

  // <div className="bx--list-box__menu">
    // {items.map((item, i) => (
      // <div
        // key={item.id}
        // className={cx({
          // 'bx--list-box__menu-item': true,
          // 'bx--list-box__menu-item--highlighted': highlightedIndex === i,
          // 'bx--list-box__menu-item--active': selectedItem.indexOf(item) !== -1,
        // })}
        // {...getItemProps({
          // item,
          // index: i,
        // })}>
        // <CheckboxReadOnly
          // id={item.id}
          // isChecked={selectedItem.indexOf(item) !== -1}
          // label={itemToString(item)}
          // name={itemToString(item)}
        // />
      // </div>
    // ))}
  // </div>
export default ListBoxMenu;
