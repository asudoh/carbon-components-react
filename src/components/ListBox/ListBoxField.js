import React from 'react';
import PropTypes from 'prop-types';
import ListBoxMenuIcon from './ListBoxMenuIcon';
import ListBoxSelection from './ListBoxSelection';
import childrenOf from '../../prop-types/childrenOf';

/**
 * `ListBoxField` is responsible for creating the containing node for valid
 * elements inside of a field. It also provides a11y-related attributes like
 * `role` to make sure a user can focus the given field.
 */
const ListBoxField = ({ children, prefix, ...rest }) => (
  <div
    role="button"
    className={`${prefix}--list-box__field`}
    tabIndex="0"
    {...rest}>
    {children}
  </div>
);

ListBoxField.propTypes = {
  /**
   * Provide the contents of your ListBoxField
   */
  children: childrenOf([ListBoxMenuIcon, ListBoxSelection, 'span', 'input']),

  /**
   * The selector prefix
   */
  prefix: PropTypes.string,
};

ListBoxField.defaultProps = {
  prefix: 'bx',
};

export default ListBoxField;
