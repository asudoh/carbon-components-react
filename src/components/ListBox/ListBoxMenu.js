import React from 'react';
import PropTypes from 'prop-types';
import ListBoxMenuItem from './ListBoxMenuItem';
import childrenOfType from '../../prop-types/childrenOfType';

/**
 * `ListBoxMenu` is a simple container node that isolates the `list-box__menu`
 * class into a single component. It is also being used to validate given
 * `children` components.
 */
const ListBoxMenu = ({ children, prefix, ...rest }) => {
  return (
    <div className={`${prefix}--list-box__menu`} {...rest}>
      {children}
    </div>
  );
};

ListBoxMenu.propTypes = {
  /**
   * Provide the contents of your ListBoxMenu
   */
  children: childrenOfType(ListBoxMenuItem),

  /**
   * The selector prefix
   */
  prefix: PropTypes.string,
};

ListBoxMenu.defaultProps = {
  prefix: 'bx',
};

export default ListBoxMenu;
