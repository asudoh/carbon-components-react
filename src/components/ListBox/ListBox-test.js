import React from 'react';
import { shallow } from 'enzyme';
import ListBox from '../ListBox';
import ListBoxMenu from '../ListBoxMenu';
import { generateItems, generateGenericItem } from '../../tools/testing/items';

describe('ListBox', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      items: generateItems(5, generateGenericItem),
      label: 'Label',
      selectedItem: [],
      isOpen: false,
      itemToString: jest.fn(({ label }) => label),
      clearSelection: jest.fn(),
      onToggleMenu: jest.fn(),
      getItemProps: jest.fn(),
      getButtonProps: jest.fn(),
      innerRef: jest.fn(),
    };
  });

  it('should render', () => {
    const wrapper = shallow(<ListBox {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call `getButtonProps` to set downshift props', () => {
    const wrapper = shallow(<ListBox {...mockProps} />);
    expect(mockProps.getButtonProps).toHaveBeenCalledWith({
      onClick: mockProps.onToggleMenu,
    });
  });

  it('should display a `ListBoxMenu` if `isOpen` is true', () => {
    const wrapper = shallow(<ListBox {...mockProps} isOpen={true} />);
    expect(wrapper.find(ListBoxMenu).length).toBe(1);
  });

  it('should set an inline class if `type="inline"`', () => {
    const wrapper = shallow(<ListBox {...mockProps} type="inline" />);
    expect(wrapper.find('.bx--list-box--inline').length).toBe(1);
  });
});
