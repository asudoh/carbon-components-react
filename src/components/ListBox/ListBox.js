import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import ListBoxMenu from '../ListBoxMenu';
import ListBoxField from '../ListBoxField';

const ListBox = ({
  type = 'default',
  items,
  selectedItem,
  label,
  clearSelection,
  isOpen,
  onToggleMenu,
  itemToString,
  highlightedIndex,
  getItemProps,
  getButtonProps,
  innerRef,
}) => {
  const className = cx({
    'bx--list-box': true,
    'bx--list-box--inline': type === 'inline',
  });

  return (
    <div className={className} ref={innerRef}>
      <ListBoxField
        type={type}
        label={label}
        selectedItem={selectedItem}
        clearSelection={clearSelection}
        isOpen={isOpen}
        {...getButtonProps({ onClick: onToggleMenu })}
      />
      {isOpen && (
        <ListBoxMenu
          items={items}
          selectedItem={selectedItem}
          itemToString={itemToString}
          getItemProps={getItemProps}
          highlightedIndex={highlightedIndex}
        />
      )}
    </div>
  );
};

ListBox.propTypes = {
  type: PropTypes.oneOf(['default', 'inline']),
  items: PropTypes.array.isRequired,
  label: PropTypes.string.isRequired,
  selectedItem: PropTypes.array.isRequired,
  isOpen: PropTypes.bool.isRequired,
  itemToString: PropTypes.func.isRequired,
  clearSelection: PropTypes.func.isRequired,
  onToggleMenu: PropTypes.func.isRequired,
  getItemProps: PropTypes.func.isRequired,
  getButtonProps: PropTypes.func.isRequired,
  innerRef: PropTypes.func.isRequired,
};

export default ListBox;
