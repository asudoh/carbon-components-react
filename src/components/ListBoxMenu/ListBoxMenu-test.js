import React from 'react';
import { shallow } from 'enzyme';
import ListBoxMenu from '../ListBoxMenu';
import {
  generateItems,
  generateGenericItem,
} from '../../tools/testing/items';

describe('ListBoxMenu', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      itemToString: jest.fn(({ label }) => label),
      getItemProps: jest.fn(() => ({})),
    };
  });

  it('should render', () => {
    const wrapper = shallow(
      <ListBoxMenu
        items={generateItems(5, generateGenericItem)}
        selectedItem={[]}
        highlightedIndex={null}
        {...mockProps}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should call `getItemProps` for each item that is generated', () => {
    const itemsCount = 5;
    shallow(
      <ListBoxMenu
        items={generateItems(itemsCount, generateGenericItem)}
        selectedItem={[]}
        highlightedIndex={null}
        {...mockProps}
      />
    );
    expect(mockProps.getItemProps).toHaveBeenCalledTimes(itemsCount);
  });

  describe('with selected items', () => {
    it('should add an active class to selected items', () => {
      const items = generateItems(5, generateGenericItem);
      const wrapper = shallow(
        <ListBoxMenu
          items={items}
          selectedItem={[items[0], items[1]]}
          highlightedIndex={null}
          {...mockProps}
        />
      );
      expect(wrapper.find('.bx--list-box__menu-item--active').length).toBe(2);
    });

    it('should set the checkbox value to checked for a selected item', () => {
      const items = generateItems(5, generateGenericItem);
      const wrapper = shallow(
        <ListBoxMenu
          items={items}
          selectedItem={[items[0], items[1]]}
          highlightedIndex={null}
          {...mockProps}
        />
      );
      wrapper.find('.bx--list-box__menu-item--active input').forEach(node => {
        expect(node.prop('checked')).toBe(true);
      });
    });
  });

  describe('with highlighted index', () => {
    it('should add a highlighted class for the specific item at the index', () => {
      const wrapper = shallow(
        <ListBoxMenu
          items={generateItems(5, generateGenericItem)}
          selectedItem={[]}
          highlightedIndex={0}
          {...mockProps}
        />
      );
      expect(wrapper.find('.bx--list-box__menu-item--highlighted').length).toBe(
        1
      );
    });
  });
});
