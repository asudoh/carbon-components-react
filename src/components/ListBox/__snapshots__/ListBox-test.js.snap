// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBox should render 1`] = `
<div
  className="bx--list-box"
>
  <ListBoxField
    clearSelection={[Function]}
    isOpen={false}
    label="Label"
    selectedItem={Array []}
    type="default"
  />
</div>
`;
